ygot/render.go

```diff
@@ -1203,45 +1233,90 @@ func structJSON(s GoStruct, parentMod string, args jsonOutputConfig) (map[string
 			continue
 		}
 
-		if prependmods != nil && len(mapPaths) != len(prependmods) {
-			errs.Add(fmt.Errorf("%s: number of paths and modules in struct tag not the same: (paths: %v, modules: %v)", fType.Name, len(mapPaths), len(prependmods)))
-			continue
-		}
+		// This is a shortcut to avoid additional allocations and improve performance.
+		//
+		// The structTagToLibPaths call performs many allocations which can be prevented
+		// by using this shortcut.
+		//
+		// Currently, this shortcut only handles cases when not prepending the module name
+		// and not having a "|" separator in the path tag.
+		if prependmods == nil && !strings.Contains(pathAnnotation, "|") {
+			parent := jsonout
 
-		for i, p := range mapPaths {
-			if prependmods != nil && p.Len() != len(prependmods[i]) {
-				errs.Add(fmt.Errorf("number of paths and modules elements not the same: (paths: %v, modules: %v)", p, prependmods[i]))
+			if !strings.Contains(pathAnnotation, "/") {
+				parent[pathAnnotation] = value
 				continue
 			}
 
-			parent := jsonout
-			j := 0
-			for ; j != p.Len()-1; j++ {
-				k, err := p.StringElemAt(j)
-				if err != nil {
-					errs.Add(err)
+			pathSplit := strings.Split(pathAnnotation, "/")
+			for i, p := range pathSplit {
+				if p == "" {
 					continue
 				}
 
-				if prependmods != nil && prependmods[i][j] != "" {
-					k = fmt.Sprintf("%s:%s", prependmods[i][j], k)
-				}
-
-				if _, ok := parent[k]; !ok {
-					parent[k] = map[string]interface{}{}
+				if i < len(pathSplit)-1 {
+					if _, ok := parent[p]; !ok {
+						parent[p] = map[string]interface{}{}
+					}
+					parent = parent[p].(map[string]interface{})
+				} else {
+					parent[p] = value
 				}
-				parent = parent[k].(map[string]interface{})
 			}
-			k, err := p.LastStringElem()
+
+			continue
+		}
+
+		func() {
+			mapPaths := mapPathsPool.Get().([]*gnmiPath)
+			defer mapPathsPool.Put(mapPaths)
```

Comment:
It would be nice if this can be refactored into its own function now that there are more optimizations and code added.

Reply:
Thanks a lot for all the comments and notes. I agree that some of them seem more like trading complexity for small gains now (many are side changes added while hunting for big performance improvements for our particular use case). The use of `sync.Pool` here isn't justified by significant GC pressure reduction as well. I'll read through all of your comments again and I think we can probably revert the use of `sync.Pool`, that way this PR can be much simplified and I think it can focus on improving `writeIETFScalarJSON` based on your comment.